x/ccv/provider/types/keys.go

```diff
@@ -275,38 +275,41 @@ func ParsePendingPacketDataKey(key []byte) (string, uint64, error) {
 }
 
 // PendingSlashPacketEntryKey returns the key for storing a pending slash packet entry.
-//
-// Note: It's not expected for a single consumer chain to send a slash packet for the same validator more than once
-// in the same block. Hence why this key should ber unique per slash packet. However, if a malicious consumer did send
-// duplicate slash packets in the same block, the slash packet entry would simply be overwritten.
 func PendingSlashPacketEntryKey(packetEntry SlashPacketEntry) []byte {
 	timeBz := sdk.FormatTimeBytes(packetEntry.RecvTime)
 	timeBzL := len(timeBz)
 	return AppendMany(
 		[]byte{PendingSlashPacketEntryBytePrefix},
 		sdk.Uint64ToBigEndian(uint64(timeBzL)),
 		timeBz,
-		HashBytes(packetEntry.ValAddr),
+		sdk.Uint64ToBigEndian(packetEntry.IbcSeqNum),
 		[]byte(packetEntry.ConsumerChainID),
 	)
 }
 
 // ParsePendingSlashPacketEntryKey returns the received time and chainID for a pending slash packet entry key
-func ParsePendingSlashPacketEntryKey(bz []byte) (time.Time, string) {
```

Comment:
Can you explain more about why this needed to be changed?

Reply:
There's 2 queues:
1. global queue holding SlashPacketEntry
2. per-chain queue holding both VSCMaturedPacketData and SlashPacketData

When querying the global queue, the data in it has no value to validators. SlashPacketEntry is just this:
```golang

// A persisted queue entry indicating that a slash packet data instance needs to be handled
type SlashPacketEntry struct {
	RecvTime time.Time

	ConsumerChainID string

	IbcSeqNum uint64

	ValAddr []byte
}
```
It was brought to my attention that this data structure is not informative enough for the normal user since it is used for internal processing. Instead, it is better to provide the actual slash packet data (as transferred over IBC) since it contains useful information relevant to the user.

```golang
type SlashPacketData struct {
	Validator types.Validator
	ValsetUpdateId uint64
	Infraction types1.InfractionType
}
```

`SlashPacketData` is not available in the global queue -> we need to get it from the per-chain queue.

It was not possible to get it from the per-chain queue without adding the newer changes from `circuit-breaker` branch. The changes in question expand the per-chain queue key and add `IbcSeq` to `SlashPacketEntry`. With all data readily available we can fetch the exact slash packet that's waiting to be processed (validator, infraction, valsetUpdateId) and enrich the query output

TL;DR:
In the `goc-december` version of throttling we did not save the IBC sequence number inside the `SlashPacketEntry` in the global queue, so we could not connect the `SlashPacketEntry` to the per-chain `SlashPacketData`